src/coreclr/tools/Common/TypeSystem/IL/Stubs/AsyncResumptionStub.cs

```diff
@@ -32,22 +32,119 @@ public AsyncResumptionStub(MethodDesc owningMethod)
+        private MethodIL InitializeMethodIL()
         {
-            var emitter = new ILEmitter();
-            ILCodeStream codeStream = emitter.NewCodeStream();
+            Debug.Assert(_methodIL == null);
+            ILEmitter ilEmitter = new ILEmitter();
+            ILCodeStream ilStream = ilEmitter.NewCodeStream();
+
+            // Ported from jitinterface.cpp CEEJitInfo::getAsyncResumptionStub
+            // Emitted IL:
+            // if (!_owningMethod.Signature.IsStatic)
+            // {
+            //     if (_owningMethod.OwningType.IsValueType)
+            //         ldc.i4.0
+            //         conv.u
+            //     else
+            //         ldnull
+            // }
+            // foreach (param in _owningMethod.Signature)
+            // {
+            //     ldloca.s <local>
+            //     initobj <param type>
+            //     ldloc.s <local>
+            // }
+            // ldftn <_owningMethod>
+            // calli <this.Signature>
+            // if (!returnsVoid)
+            //     stloc.s <resultLocal>
+            // call System.StubHelpers.StubHelpers::AsyncCallContinuation()
+            // stloc.s <newContinuationLocal>
+            // if (!returnsVoid)
+            // {
+            //     ldloca.s <newContinuationLocal>
+            //     brtrue.s done_result
+            //     ldarg.1
+            //     ldloc.s <resultLocal>
+            //     stobj <return type>
+            //   done_result:
+            // }
+            // ldloc.s <newContinuationLocal>
+            // ret
+
+            // if it has this pointer
+            if (!_owningMethod.Signature.IsStatic)
+            {
+                if (_owningMethod.OwningType.IsValueType)
+                {
+                    ilStream.EmitLdc(0);
+                    ilStream.Emit(ILOpcode.conv_u);
+                }
+                else
+                {
+                    ilStream.Emit(ILOpcode.ldnull);
+                }
+            }
+
+            foreach (var param in _owningMethod.Signature)
+            {
+                var local = ilEmitter.NewLocal(param);
+                ilStream.EmitLdLoca(local);
+                ilStream.Emit(ILOpcode.initobj, ilEmitter.NewToken(param));
+                ilStream.EmitLdLoc(local);
+            }
+            ilStream.Emit(ILOpcode.ldftn, ilEmitter.NewToken(_owningMethod));
+            ilStream.Emit(ILOpcode.calli, ilEmitter.NewToken(this.Signature));
+
+            bool returnsVoid = _owningMethod.Signature.ReturnType != Context.GetWellKnownType(WellKnownType.Void);
+            Internal.IL.Stubs.ILLocalVariable resultLocal = default;
+            if (!returnsVoid)
+            {
+                resultLocal = ilEmitter.NewLocal(_owningMethod.Signature.ReturnType);
+                ilStream.EmitStLoc(resultLocal);
+            }
+
+            MethodDesc asyncCallContinuation = Context.SystemModule.GetKnownType("System.StubHelpers"u8, "StubHelpers"u8)
```

Comment:
Can we move this to AsyncHelpers so that we do not have to add StubHelpers.cs to NativeAOT? 

StubHelpers in CoreCLR is everything-in-the-kitchen-sink type. It is not pretty.